ts/sagas/startup/generateCryptoKeyPair.ts

```diff
@@ -1,44 +1,106 @@
+import { call, select } from "typed-redux-saga/macro";
+import { isLollipopEnabledSelector } from "../../store/reducers/backendStatus";
 import {
   checkPublicKeyExists,
-  isKeyAlreadyGenerated,
-  setKeyAlreadyGenerated
+  setKeyGenerationInfo,
+  getKeyGenerationInfo,
+  KeyGenerationInfo
 } from "../../utils/crypto";
 import { mixpanelTrack } from "./../../mixpanel";
 
-const KEY_NAME = "lp-temp-key";
+/**
+ * Generates a new crypto key pair.
+ */
+export function* cryptoKeyGenerationSaga(
+  keyTag: string,
+  previousKeyTag?: string
+) {
+  const isLollipopEnabled = yield* select(isLollipopEnabledSelector);
+  if (isLollipopEnabled) {
+    // Every new login we need to regenerate a brand new key pair.
+    yield* deletePreviousCryptoKeyPair(previousKeyTag);
+    yield* call(generateCryptoKeyPair, keyTag);
+  }
+}
 
-export function* generateCryptoKeyPair() {
-  try {
-    const keyAlreadyExistsOnKeystore = yield* call(
-      checkPublicKeyExists,
-      KEY_NAME
-    );
-    const keyHasBeenAlreadyGenerated = yield* call(
-      isKeyAlreadyGenerated,
-      KEY_NAME
-    );
-    const shouldWeGenerateKey = !(
-      keyHasBeenAlreadyGenerated || keyAlreadyExistsOnKeystore
-    );
-
-    if (shouldWeGenerateKey) {
-      const key = yield* call(generate, KEY_NAME);
-      void mixpanelTrack("LOLLIPOP_KEY_GENERATION_SUCCESS", {
-        kty: key.kty
-      });
-      yield* call(setKeyAlreadyGenerated, KEY_NAME, key.kty);
-      yield* call(deleteKey, KEY_NAME);
-    }
-  } catch (e) {
-    const errorMessage = (e as CryptoError).message;
+/**
+ * Sends to mixpanel the crypto key pair generation events.
+ */
+export function* trackMixpanelCryptoKeyPairEvents(keyTag: string) {
+  const keyInfo = yield* call(getKeyGenerationInfo, keyTag);
+
+  if (keyInfo && !keyInfo.errorCode) {
+    void mixpanelTrack("LOLLIPOP_KEY_GENERATION_SUCCESS", {
+      kty: keyInfo.keyType
+    });
+  }
+
+  if (keyInfo && keyInfo.errorCode) {
     void mixpanelTrack("LOLLIPOP_KEY_GENERATION_FAILURE", {
-      reason: errorMessage
+      reason: keyInfo.errorCode,
+      resonMoreInfo: keyInfo.userInfo
     });
   }
 }
+
+/**
+ * Deletes a previous saved crypto key pair.
+ */
+export function* deletePreviousCryptoKeyPair(keyTag?: string) {
+  if (keyTag) {
+    yield* deleteCryptoKeyPair(keyTag);
+  }
+}
+
+/**
+ * Deletes the crypto key pair corresponding to the provided `keyTag`.
+ */
+function* deleteCryptoKeyPair(keyTag: string) {
+  // Key is persisted even after uninstalling the application on iOS.
+  const keyAlreadyExistsOnKeystore = yield* call(checkPublicKeyExists, keyTag);
+
+  if (keyAlreadyExistsOnKeystore) {
+    try {
+      yield* call(deleteKey, keyTag);
+    } catch (e) {
+      yield* saveKeyGenerationFailureInfo(keyTag, e);
+    }
+  }
+}
+
+/**
+ * Generates a new crypto key pair.
+ */
+function* generateCryptoKeyPair(keyTag: string) {
+  try {
+    // Remove an already existing key with the same tag.
+    deleteCryptoKeyPair(keyTag);
```

Comment:
if we delete the `keyTag` inside `cryptoKeyGenerationSaga`, is it necessary to repeat the operation?

Reply:
In the saga we remove a previous generated key (if any) while in `generateCryptoKeyPair` we delete a key with the same `keyTag` we are actually using to crate the new key pair if, for some reason, something happened (a remote possibility) that put us in this edge case.